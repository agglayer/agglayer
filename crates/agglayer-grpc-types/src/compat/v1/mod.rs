// Helper macro used by the rest of this module
macro_rules! required_field {
    ($from:expr, $field:ident) => {
        $from
            .$field
            .ok_or(Error::missing_field(stringify!($field)))?
            .try_into()
            .map_err(|e: Error| e.inside_field(stringify!($field)))?
    };
}

mod certificate;
mod certificate_header;
mod certificate_id;
mod epoch_configuration;
mod error_kinds;

pub use agglayer_interop::grpc::compat::v1::Error;
pub use agglayer_interop::grpc::compat::v1::ErrorKind;

#[cfg(test)]
pub mod tests;
