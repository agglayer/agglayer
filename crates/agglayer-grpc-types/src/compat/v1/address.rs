use agglayer_types::Address;
use prost::bytes::Bytes;

use super::Error;
use crate::protocol::types::v1;

impl TryFrom<v1::FixedBytes20> for Address {
    type Error = Error;

    fn try_from(value: v1::FixedBytes20) -> Result<Self, Self::Error> {
        Ok(Address::from(<[u8; 20]>::try_from(value)?))
    }
}

impl From<Address> for v1::FixedBytes20 {
    fn from(value: Address) -> Self {
        v1::FixedBytes20 {
            value: Bytes::copy_from_slice(&value.0 .0),
        }
    }
}
