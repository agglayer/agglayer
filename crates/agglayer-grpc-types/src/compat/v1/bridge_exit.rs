use agglayer_types::{BridgeExit, LeafType, NetworkId};

use super::Error;
use crate::protocol::types::v1;

impl TryFrom<v1::BridgeExit> for BridgeExit {
    type Error = Error;

    fn try_from(value: v1::BridgeExit) -> Result<Self, Self::Error> {
        Ok(BridgeExit {
            leaf_type: match value.leaf_type() {
                v1::LeafType::Transfer => LeafType::Transfer,
                v1::LeafType::Message => LeafType::Message,
                _ => {
                    let t = value.leaf_type;
                    return Err(Error::invalid_data(format!("invalid leaf type: {t}")));
                }
            },
            token_info: required_field!(value, token_info),
            dest_network: NetworkId::new(value.dest_network),
            dest_address: required_field!(value, dest_address),
            amount: required_field!(value, amount),
            metadata: value
                .metadata
                .map(TryInto::try_into)
                .transpose()
                .map_err(|e: Error| e.inside_field("metadata"))?,
        })
    }
}

impl From<BridgeExit> for v1::BridgeExit {
    fn from(value: BridgeExit) -> Self {
        v1::BridgeExit {
            leaf_type: match value.leaf_type {
                LeafType::Transfer => v1::LeafType::Transfer,
                LeafType::Message => v1::LeafType::Message,
            }
            .into(),
            token_info: Some(value.token_info.into()),
            dest_network: value.dest_network.into(),
            dest_address: Some(value.dest_address.into()),
            amount: Some(value.amount.into()),
            metadata: value.metadata.map(Into::into),
        }
    }
}
