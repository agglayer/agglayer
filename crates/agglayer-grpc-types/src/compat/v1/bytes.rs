use super::Error;
use crate::protocol::types::v1;

impl TryFrom<v1::FixedBytes20> for [u8; 20] {
    type Error = Error;

    fn try_from(value: v1::FixedBytes20) -> Result<Self, Self::Error> {
        (&*value.value)
            .try_into()
            .map_err(|_| Error::WrongBytesLength {
                expected: 20,
                actual: value.value.len(),
            })
    }
}

impl TryFrom<v1::FixedBytes32> for [u8; 32] {
    type Error = Error;

    fn try_from(value: v1::FixedBytes32) -> Result<Self, Self::Error> {
        (&*value.value)
            .try_into()
            .map_err(|_| Error::WrongBytesLength {
                expected: 32,
                actual: value.value.len(),
            })
    }
}
