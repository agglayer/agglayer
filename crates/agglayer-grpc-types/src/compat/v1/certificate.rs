use agglayer_types::{Certificate, NetworkId};

use super::Error;
use crate::node::types::v1;

impl TryFrom<v1::Certificate> for Certificate {
    type Error = Error;

    fn try_from(value: v1::Certificate) -> Result<Self, Self::Error> {
        Ok(Certificate {
            network_id: NetworkId::new(value.network_id),
            height: value.height,
            prev_local_exit_root: required_field!(value, prev_local_exit_root),
            new_local_exit_root: required_field!(value, new_local_exit_root),
            bridge_exits: value
                .bridge_exits
                .into_iter()
                .map(TryInto::try_into)
                .collect::<Result<_, _>>()
                .map_err(|e: Error| e.inside_field("bridge_exits"))?,
            imported_bridge_exits: value
                .imported_bridge_exits
                .into_iter()
                .map(TryInto::try_into)
                .collect::<Result<_, _>>()
                .map_err(|e: Error| e.inside_field("imported_bridge_exits"))?,
            aggchain_data: required_field!(value, aggchain_data),
            metadata: required_field!(value, metadata),
            custom_chain_data: value.custom_chain_data.to_vec(),
        })
    }
}

impl TryFrom<Certificate> for v1::Certificate {
    type Error = Error;

    fn try_from(value: Certificate) -> Result<Self, Self::Error> {
        Ok(v1::Certificate {
            network_id: value.network_id.into(),
            height: value.height,
            prev_local_exit_root: Some(value.prev_local_exit_root.into()),
            new_local_exit_root: Some(value.new_local_exit_root.into()),
            bridge_exits: value.bridge_exits.into_iter().map(Into::into).collect(),
            imported_bridge_exits: value
                .imported_bridge_exits
                .into_iter()
                .map(Into::into)
                .collect(),
            aggchain_data: Some(value.aggchain_data.try_into()?),
            metadata: Some(value.metadata.into()),
            custom_chain_data: value.custom_chain_data.into(),
        })
    }
}
