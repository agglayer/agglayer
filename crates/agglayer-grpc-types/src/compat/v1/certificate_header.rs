use agglayer_types::{CertificateHeader, CertificateStatus, CertificateStatusError};

use crate::node::types::v1;

impl From<CertificateStatusError> for v1::CertificateStatusError {
    fn from(value: CertificateStatusError) -> Self {
        v1::CertificateStatusError {
            // Display value with the whole stack trace
            message: format!("{value:?}").into_bytes().into(),
        }
    }
}

impl From<CertificateHeader> for v1::CertificateHeader {
    fn from(value: CertificateHeader) -> Self {
        let (status, error) = match value.status {
            CertificateStatus::Pending => (v1::CertificateStatus::Pending, None),
            CertificateStatus::Proven => (v1::CertificateStatus::Proven, None),
            CertificateStatus::Candidate => (v1::CertificateStatus::Candidate, None),
            CertificateStatus::InError { error } => {
                (v1::CertificateStatus::InError, Some(error.into()))
            }
            CertificateStatus::Settled => (v1::CertificateStatus::Settled, None),
        };
        v1::CertificateHeader {
            network_id: value.network_id.into(),
            height: value.height,
            epoch_number: value.epoch_number,
            certificate_index: value.certificate_index,
            certificate_id: Some(value.certificate_id.into()),
            prev_local_exit_root: Some(value.prev_local_exit_root.into()),
            new_local_exit_root: Some(value.new_local_exit_root.into()),
            metadata: Some(value.metadata.into()),
            status: status.into(),
            error,
            settlement_tx_hash: value.settlement_tx_hash.map(Into::into),
        }
    }
}
