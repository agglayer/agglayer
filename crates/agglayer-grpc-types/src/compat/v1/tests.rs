use agglayer_types::{
    aggchain_proof::AggchainData, primitives::SignatureError, Address, BridgeExit, CertificateId,
    ClaimFromMainnet, ClaimFromRollup, Digest, EpochConfiguration, GlobalIndex, ImportedBridgeExit,
    L1InfoTreeLeaf, L1InfoTreeLeafInner, MerkleProof, TokenInfo, U256,
};
use prost::Message;

use super::Error;
use crate::protocol::types::v1;

#[cfg(fuzzing)]
pub mod fuzzing_workarounds {
    // TODO: these all should be in sp1 upstream, but they're not marked as #[used]
    // and so disappear with optimizations
    #[no_mangle]
    pub extern "C" fn read_vec_raw() {
        unimplemented!("SP1 workaround, should never be called")
    }
    #[no_mangle]
    pub extern "C" fn _end() {
        unimplemented!("SP1 workaround, should never be called")
    }
    #[used]
    static _USED: [extern "C" fn(); 2] = [read_vec_raw, _end];
}

#[rstest::rstest]
#[case::error("no_proof", Error::missing_field("proof"))]
#[case::error("bad_data", Error::invalid_data("invalid value".to_owned()))]
#[case::error("bad_data_in_field", Error::invalid_data("invalid value".to_owned()).inside_field("value"))]
#[case::error("bad_data_in_nested", Error::invalid_data("invalid value".to_owned()).inside_field("value").inside_field("data"))]
#[case::error("failed_ser", Error::serializing_proof(Box::new(bincode::ErrorKind::Custom("failed".to_owned()))))]
#[case::error("failed_deser", Error::deserializing_proof(Box::new(bincode::ErrorKind::Custom("failed".to_owned()))))]
#[case::error("bad_sig", Error::parsing_signature(SignatureError::InvalidParity(5)))]
#[case::error("bad_sig_in_nested", Error::parsing_signature(SignatureError::InvalidParity(5)).inside_field("signature").inside_field("data"))]
fn error_messages(#[case] name: &str, #[case] error: Error) {
    insta::assert_snapshot!(format!("{name}/display"), error);
    insta::assert_debug_snapshot!(format!("{name}/kind"), error.kind());
    insta::assert_snapshot!(
        format!("{name}/debug"),
        format!("{:?}", anyhow::Error::from(error))
    );
}

macro_rules! make_parser_fuzzers {
    ($test:ident, $proto:ty, $type:ty) => {
        #[test]
        fn $test() {
            bolero::check!().for_each(|bytes| {
                if let Ok(proto) = <$proto>::decode(bytes) {
                    let _ = <$type>::try_from(proto);
                };
            })
        }
    };
}

make_parser_fuzzers!(fuzz_parser_address, v1::FixedBytes20, Address);
make_parser_fuzzers!(fuzz_parser_aggchain_data, v1::AggchainData, AggchainData);
make_parser_fuzzers!(fuzz_parser_bridge_exit, v1::BridgeExit, BridgeExit);
make_parser_fuzzers!(fuzz_parser_certificate_id, v1::CertificateId, CertificateId);
make_parser_fuzzers!(
    fuzz_parser_claim_from_mainnet,
    v1::ClaimFromMainnet,
    ClaimFromMainnet
);
make_parser_fuzzers!(
    fuzz_parser_claim_from_rollup,
    v1::ClaimFromRollup,
    ClaimFromRollup
);
make_parser_fuzzers!(fuzz_parser_digest, v1::FixedBytes32, Digest);
make_parser_fuzzers!(
    fuzz_parser_epoch_configuration,
    v1::EpochConfiguration,
    EpochConfiguration
);
make_parser_fuzzers!(fuzz_parser_global_index, v1::FixedBytes32, GlobalIndex);
make_parser_fuzzers!(
    fuzz_parser_imported_bridge_exit,
    v1::ImportedBridgeExit,
    ImportedBridgeExit
);
make_parser_fuzzers!(
    fuzz_parser_l1_info_tree_leaf_with_context,
    v1::L1InfoTreeLeafWithContext,
    L1InfoTreeLeaf
);
make_parser_fuzzers!(
    fuzz_parser_l1_info_tree_leaf_inner,
    v1::L1InfoTreeLeaf,
    L1InfoTreeLeafInner
);
make_parser_fuzzers!(fuzz_parser_merkle_proof, v1::MerkleProof, MerkleProof);
make_parser_fuzzers!(fuzz_parser_token_info, v1::TokenInfo, TokenInfo);
make_parser_fuzzers!(fuzz_parser_u256, v1::FixedBytes32, U256);

macro_rules! make_round_trip_fuzzers {
    ($test:ident, $proto:ty, $type:ty) => {
        #[test]
        fn $test() {
            bolero::check!()
                .with_arbitrary::<$type>()
                .for_each(|input: &$type| {
                    let proto: $proto = input.clone().into();
                    let output = <$type>::try_from(proto).unwrap();
                    assert_eq!(input, &output);
                })
        }
    };
}

make_round_trip_fuzzers!(fuzz_round_trip_address, v1::FixedBytes20, Address);
#[test]
fn fuzz_round_trip_aggchain_data() {
    bolero::check!()
        .with_arbitrary::<AggchainData>()
        .for_each(|input| {
            let proto: v1::AggchainData = input.clone().try_into().unwrap();
            let _output = AggchainData::try_from(proto).unwrap();
            // There's no good way to check equality of two stark proofs:
            // assert_eq!(input, &output);
        })
}
make_round_trip_fuzzers!(fuzz_round_trip_bridge_exit, v1::BridgeExit, BridgeExit);
make_round_trip_fuzzers!(
    fuzz_round_trip_certificate_id,
    v1::CertificateId,
    CertificateId
);
make_round_trip_fuzzers!(
    fuzz_round_trip_claim_from_mainnet,
    v1::ClaimFromMainnet,
    ClaimFromMainnet
);
make_round_trip_fuzzers!(
    fuzz_round_trip_claim_from_rollup,
    v1::ClaimFromRollup,
    ClaimFromRollup
);
make_round_trip_fuzzers!(fuzz_round_trip_digest, v1::FixedBytes32, Digest);
make_round_trip_fuzzers!(
    fuzz_round_trip_epoch_configuration,
    v1::EpochConfiguration,
    EpochConfiguration
);
make_round_trip_fuzzers!(fuzz_round_trip_global_index, v1::FixedBytes32, GlobalIndex);
make_round_trip_fuzzers!(
    fuzz_round_trip_imported_bridge_exit,
    v1::ImportedBridgeExit,
    ImportedBridgeExit
);
make_round_trip_fuzzers!(
    fuzz_round_trip_l1_info_tree_leaf_with_context,
    v1::L1InfoTreeLeafWithContext,
    L1InfoTreeLeaf
);
make_round_trip_fuzzers!(
    fuzz_round_trip_l1_info_tree_leaf_inner,
    v1::L1InfoTreeLeaf,
    L1InfoTreeLeafInner
);
make_round_trip_fuzzers!(fuzz_round_trip_merkle_proof, v1::MerkleProof, MerkleProof);
make_round_trip_fuzzers!(fuzz_round_trip_token_info, v1::TokenInfo, TokenInfo);
make_round_trip_fuzzers!(fuzz_round_trip_u256, v1::FixedBytes32, U256);
