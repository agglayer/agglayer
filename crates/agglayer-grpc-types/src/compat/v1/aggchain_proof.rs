use std::collections::HashMap;

use agglayer_types::aggchain_proof::{AggchainData, Proof};
use bincode::Options as _;
use prost::bytes::Bytes;

use super::Error;
use crate::protocol::types::v1;

fn sp1v4_bincode_options() -> impl bincode::Options {
    bincode::DefaultOptions::new()
        .with_big_endian()
        .with_fixint_encoding()
}

impl TryFrom<v1::AggchainData> for AggchainData {
    type Error = Error;

    fn try_from(value: v1::AggchainData) -> Result<Self, Self::Error> {
        Ok(match value.data {
            Some(v1::aggchain_data::Data::Signature(signature)) => AggchainData::ECDSA {
                signature: (&*signature.value)
                    .try_into()
                    .map_err(Error::ParsingSignature)?,
            },
            Some(v1::aggchain_data::Data::Generic(proof)) => AggchainData::Generic {
                aggchain_params: required_field!(proof, aggchain_params),
                proof: match proof.proof {
                    Some(v1::aggchain_proof::Proof::Sp1Stark(proof)) => Proof::SP1Stark(Box::new(
                        sp1v4_bincode_options()
                            .deserialize(&proof)
                            .map_err(Error::DeserializingProof)?,
                    )),
                    None => return Err(Error::MissingField("data.proof")),
                },
            },
            None => return Err(Error::MissingField("data")),
        })
    }
}

impl TryFrom<AggchainData> for v1::AggchainData {
    type Error = Error;

    fn try_from(value: AggchainData) -> Result<Self, Self::Error> {
        Ok(v1::AggchainData {
            data: Some(match value {
                AggchainData::ECDSA { signature } => {
                    v1::aggchain_data::Data::Signature(v1::FixedBytes65 {
                        value: Bytes::copy_from_slice(&signature.as_bytes()),
                    })
                }
                AggchainData::Generic {
                    proof,
                    aggchain_params,
                } => v1::aggchain_data::Data::Generic(v1::AggchainProof {
                    context: HashMap::new(),
                    aggchain_params: Some(aggchain_params.into()),
                    proof: Some(v1::aggchain_proof::Proof::Sp1Stark(
                        sp1v4_bincode_options()
                            .serialize(&proof)
                            .map_err(Error::SerializingProof)?
                            .into(),
                    )),
                }),
            }),
        })
    }
}
