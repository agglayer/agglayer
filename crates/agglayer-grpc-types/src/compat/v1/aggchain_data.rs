use std::collections::HashMap;

use agglayer_types::aggchain_proof::{AggchainData, Proof};
use bincode::Options as _;
use prost::bytes::Bytes;

use super::Error;
use crate::protocol::types::v1;

fn sp1v4_bincode_options() -> impl bincode::Options {
    bincode::DefaultOptions::new()
        .with_big_endian()
        .with_fixint_encoding()
}

impl TryFrom<v1::AggchainData> for AggchainData {
    type Error = Error;

    fn try_from(value: v1::AggchainData) -> Result<Self, Self::Error> {
        Ok(match value.data {
            Some(v1::aggchain_data::Data::Signature(signature)) => AggchainData::ECDSA {
                signature: (&*signature.value)
                    .try_into()
                    .map_err(Error::parsing_signature)?,
            },
            Some(v1::aggchain_data::Data::Generic(proof)) => AggchainData::Generic {
                aggchain_params: required_field!(proof, aggchain_params),
                proof: match proof.proof {
                    Some(v1::aggchain_proof::Proof::Sp1Stark(proof)) => Proof::SP1Stark(Box::new(
                        std::panic::catch_unwind(|| sp1v4_bincode_options().deserialize(&proof))
                            .map_err(|_| {
                                Error::deserializing_proof(Box::new(bincode::ErrorKind::Custom(
                                    String::from("panic"),
                                )))
                            })?
                            .map_err(Error::deserializing_proof)?,
                    )),
                    None => return Err(Error::missing_field("proof").inside_field("data")),
                },
            },
            None => return Err(Error::missing_field("data")),
        })
    }
}

impl TryFrom<AggchainData> for v1::AggchainData {
    type Error = Error;

    fn try_from(value: AggchainData) -> Result<Self, Self::Error> {
        Ok(v1::AggchainData {
            data: Some(match value {
                AggchainData::ECDSA { signature } => {
                    v1::aggchain_data::Data::Signature(v1::FixedBytes65 {
                        value: Bytes::copy_from_slice(&signature.as_bytes()),
                    })
                }
                AggchainData::Generic {
                    proof: Proof::SP1Stark(proof),
                    aggchain_params,
                } => v1::aggchain_data::Data::Generic(v1::AggchainProof {
                    context: HashMap::new(),
                    aggchain_params: Some(aggchain_params.into()),
                    proof: Some(v1::aggchain_proof::Proof::Sp1Stark(
                        sp1v4_bincode_options()
                            .serialize(&proof)
                            .map_err(Error::serializing_proof)?
                            .into(),
                    )),
                }),
            }),
        })
    }
}
