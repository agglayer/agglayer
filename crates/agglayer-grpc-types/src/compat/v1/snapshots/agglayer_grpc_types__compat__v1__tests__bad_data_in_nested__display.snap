---
source: crates/agglayer-grpc-types/src/compat/v1/tests.rs
expression: error
---
data.value: invalid value
